src/main/java/io/strimzi/kafka/quotas/ThrottleFactorSupplier.java

```diff
@@ -0,0 +1,25 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+
+package io.strimzi.kafka.quotas;
+
+import java.util.Collection;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+/**
+ * Determines the current restriction factor to be applied to the client quota.
+ * Values returned are required to be in the range <code>0.0..1.0</code> inclusive.
+ * Where a value of `1.0` implies no additional restriction over and above the defined quota.
+ * A value of `0.0` implies that there is no quota available regardless of the defined quota.
+ */
+public interface ThrottleFactorSupplier extends Supplier<Double>, Consumer<Collection<Volume>> {
+    /**
+     * Register a listener to be invoked when the throttle factor is changed.
+     * @param listener to be executed whenever the factor is updated.
+     */
+    void addUpdateListener(Runnable listener);
```

Comment:
We also need to support the static quotas for non-PRODUCE quota types, so we could have an awkward split where the non-PRODUCE quota types are handled by the existing set and PRODUCE quota types have to delegate to the throttle supplier. It might be simpler as-is to have it tell the `StaticQuotaCallback` that PRODUCE has changed so they can share the same mechanism. Though as you said the details could be cleaned up.

I also like how currently `StaticQuotaCallback` contains the knowledge about this `quotaResetRequired` API (along with it's performance requirements). Having an atomic boolean in the `ThrottleFactorSupplier` somewhere that is polled by  StaticQuotaCallback using `getAndSet(false)` implies that the `ThrottleFactorSupplier` would be aware of this quota-reset API. Maybe that's okay, if we think it makes things more readable the `ThrottleFactorSupplier` could implement `public boolean quotaResetRequired(ClientQuotaType quotaType)` but it becomes less about naively handing off the factor and more coupled to the callback API.

Another note is we went with the `setListener` style since we've ended up with two ThrottleFactorPolicy implementations and wanted a common way to wire in the listener (since telling StaticQuotaCallback was a requirement for both). A setter was something we could put in the interface. Maybe we can refactor it further so there is a single class responsible for knowing the factor has changed and telling the StaticQuotaCallback. I imagine we can get it to the point where we use the constructor to pass a single listener instead of using this setter. We could introduce an interface instead of Runnable to document that it shouldn't block up the world since it's called from the admin thread.

Reply:
@robobario and I think 0e91ea08b8ae84a0ab1fcfa218097d69eb4eed3d addresses the spirit of this conversation.